### PR TITLE
[SPARK-41848][CORE] Fixing task over-scheduled with TaskResourceProfile

### PR DESCRIPTION
**Pull request by ivoson**
### What changes were proposed in this pull request?
As described in [SPARK-41848](https://issues.apache.org/jira/browse/SPARK-41848), we update the [executor's free cores](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala#L424) based on executor's resource profile. For tasks with TaskResourceProfile in standalone cluster, since executors with default resource profile can be reused across different TaskResourceProfiles, and the task cpus can be different from the value we get from executor's resource profile.

The changes to fix the issues:
1. Decrease executor free cores when launch tasks by taskCpus from `TaskDescription`;
2. Adding taskCpus to `StatusUpdate` as well as the other resources, so that taskCpus can be reported when task is finished and we can increase executor free cores by the reported taskCpus;

### Why are the changes needed?
Fixing the bug as described in [SPARK-41848](https://issues.apache.org/jira/browse/SPARK-41848)

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
New UTs added.

### CODE REVIEW

**Comment by ivoson** [comment]
cc @Ngone51 could you please take a look at this PR? Thanks

File: core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala [hunk 1]
```diff
@@ -262,9 +269,11 @@ private[spark] class CoarseGrainedExecutorBackend(
 
   override def statusUpdate(taskId: Long, state: TaskState, data: ByteBuffer): Unit = {
     val resources = taskResources.getOrElse(taskId, Map.empty[String, ResourceInformation])
-    val msg = StatusUpdate(executorId, taskId, state, data, resources)
+    val cpus = taskCpus.getOrElse(taskId, 0)
```
Comment (Ngone51) [inline core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala:272]: Could we get the taskCpu by `executor.runningTask(taskId).taskDescription.cpus` to get rid of `taskCpus`?

File: core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala [hunk 2]
```diff
@@ -745,6 +740,12 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
     executorDataMap.get(executorId).map(_.resourcesInfo).getOrElse(Map.empty)
   }
 
+  // this function is for testing only
+  def getExecutorAvailableCpus(
```
Comment (Ngone51) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:744]: Could we make it private spark?

File: core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala [hunk 3]
```diff
@@ -418,9 +415,7 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
           val executorData = executorDataMap(task.executorId)
           // Do resources allocation here. The allocated resources will get released after the task
           // finishes.
-          val rpId = executorData.resourceProfileId
-          val prof = scheduler.sc.resourceProfileManager.resourceProfileFromId(rpId)
-          val taskCpus = ResourceProfile.getTaskCpusOrDefaultForProfile(prof, conf)
+          val taskCpus = task.cpus
           executorData.freeCores -= taskCpus
```
Comment (Ngone51) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:419]:
```suggestion
          executorData.freeCores -= task.cpus
```

File: core/src/test/scala/org/apache/spark/scheduler/CoarseGrainedSchedulerBackendSuite.scala [hunk 4]
```diff
@@ -403,6 +405,92 @@ class CoarseGrainedSchedulerBackendSuite extends SparkFunSuite with LocalSparkCo
...
   }
 
+  test("SPARK-41848: executor core decrease should base on taskCpus") {
+    val testStartTime = System.currentTimeMillis()
+
+    val execCores = 3
+    val conf = new SparkConf()
+      .set(EXECUTOR_CORES, execCores)
+      .set(SCHEDULER_REVIVE_INTERVAL.key, "1m") // don't let it auto revive during test
+      .set(EXECUTOR_INSTANCES, 0) // avoid errors about duplicate executor registrations
```
Comment (Ngone51) [inline core/src/test/scala/org/apache/spark/scheduler/CoarseGrainedSchedulerBackendSuite.scala:415]: What do you mean by "duplicate executor registrations"?

File: core/src/test/scala/org/apache/spark/scheduler/CoarseGrainedSchedulerBackendSuite.scala [hunk 5]
```diff
@@ -403,6 +405,92 @@ class CoarseGrainedSchedulerBackendSuite extends SparkFunSuite with LocalSparkCo
...
+
+    val backend = sc.schedulerBackend.asInstanceOf[TestCoarseGrainedSchedulerBackend]
+    // Request execs in the default profile.
+    backend.requestExecutors(1)
+    val mockEndpointRef = mock[RpcEndpointRef]
+    val mockAddress = mock[RpcAddress]
+    when(mockEndpointRef.send(LaunchTask)).thenAnswer((_: InvocationOnMock) => {})
+
+    var executorAddedCount: Int = 0
+    val infos = scala.collection.mutable.ArrayBuffer[ExecutorInfo]()
```
Comment (Ngone51) [inline core/src/test/scala/org/apache/spark/scheduler/CoarseGrainedSchedulerBackendSuite.scala:430]: Could you move `scala.collection.mutable` to import list?

**Comment by mridulm** [comment]
I will need to test this and revisit the code to understand the issue better.
I would expect this info to be at driver already - not sure why executors are having to let the driver know.

File: core/src/test/scala/org/apache/spark/scheduler/CoarseGrainedSchedulerBackendSuite.scala [hunk 6]
```diff
@@ -403,6 +405,92 @@ class CoarseGrainedSchedulerBackendSuite extends SparkFunSuite with LocalSparkCo
...
+    when(ts.resourceOffers(any[IndexedSeq[WorkerOffer]], any[Boolean])).thenReturn(taskDescs)
+
+    backend.driverEndpoint.send(ReviveOffers)
+
+    eventually(timeout(5 seconds)) {
+      assert(backend.getExecutorAvailableCpus("1").contains(1))
+    }
+
+    // To avoid allocating any resources immediately after releasing the resource from the task to
+    // make sure that `availableAddrs` below won't change
```
Comment (Ngone51) [inline core/src/test/scala/org/apache/spark/scheduler/CoarseGrainedSchedulerBackendSuite.scala:473]: available cpus?

**Comment by Ngone51** [comment]
@mridulm Please note that this is a new bug introduced by a recent PR (https://github.com/apache/spark/pull/37268) that hasn't been rollout.

**Comment by mridulm** [comment]
Ah ! Thanks for the context @Ngone51 - just checked git blame to see it is part of 3.4 :-)

**Comment by tgravescs** [comment]
Overall approach looks good, I think @Ngone51 covered comments I had.

File: core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala (hunk 1 above)
Comment (ivoson) [reply core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala:272]: Thanks, that looks better. Making the change.

File: core/src/test/scala/org/apache/spark/scheduler/CoarseGrainedSchedulerBackendSuite.scala (hunk 6 above)
Comment (ivoson) [reply core/src/test/scala/org/apache/spark/scheduler/CoarseGrainedSchedulerBackendSuite.scala]: Thanks, fixed.

File: core/src/test/scala/org/apache/spark/scheduler/CoarseGrainedSchedulerBackendSuite.scala (hunk 5 above)
Comment (ivoson) [reply core/src/test/scala/org/apache/spark/scheduler/CoarseGrainedSchedulerBackendSuite.scala]: done

File: core/src/test/scala/org/apache/spark/scheduler/CoarseGrainedSchedulerBackendSuite.scala (hunk 4 above)
Comment (ivoson) [reply core/src/test/scala/org/apache/spark/scheduler/CoarseGrainedSchedulerBackendSuite.scala]: Kept the comments from other UTs, while I checked the code and didn't see any chance for the duplicate registrations. Removing the comments here.

File: core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala (hunk 2 above)
Comment (ivoson) [reply core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala]: done.

File: core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala (hunk 1 above)
Comment (ivoson) [reply core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala]: done

**Review by dongjoon-hyun — COMMENTED** [review]
cc @xinrong-meng (Apache Spark 3.4 release manager) since this is filed as a blocker-level JIRA issue.

**Review by Ngone51 — APPROVED** [review]

File: core/src/test/scala/org/apache/spark/scheduler/CoarseGrainedSchedulerBackendSuite.scala [hunk 7]
```diff
@@ -403,6 +405,92 @@ class CoarseGrainedSchedulerBackendSuite extends SparkFunSuite with LocalSparkCo
       "Our unexpected executor does not have a request time.")
   }
 
+  test("SPARK-41848: executor core decrease should base on taskCpus") {
```
Comment (dongjoon-hyun) [inline core/src/test/scala/org/apache/spark/scheduler/CoarseGrainedSchedulerBackendSuite.scala:408]: nit. `decrease should base on` -> `should be decreased based on`?

**Review by dongjoon-hyun — APPROVED** [review]

File: core/src/test/scala/org/apache/spark/scheduler/CoarseGrainedSchedulerBackendSuite.scala (hunk 7 above)
Comment (ivoson) [reply core/src/test/scala/org/apache/spark/scheduler/CoarseGrainedSchedulerBackendSuite.scala]: Thanks, updated.

**Review by WeichenXu123 — APPROVED** [review]
LGTM

**Comment by dongjoon-hyun** [comment]
Merged to master. Thank you, @ivoson , @Ngone51 , @mridulm , @tgravescs , @WeichenXu123 .
Also, cc @sunchao

**Comment by dongjoon-hyun** [comment]
Sorry for missing that, @mridulm .

**Comment by mridulm** [comment]
Oh no, you are good @dongjoon-hyun - I was not actively reviewing this PR !

**Comment by Ngone51** [comment]
@mridulm That might be an alternative. I was thinking about it too. But checking the code, I found that a `TaskSet` might have been cleaned up when a task finishes, e.g., in the case of executor lost comes before the `StatusUpdate`. `TaskSchedulerImpl.statusUpdate()` has also considered this case: 
https://github.com/apache/spark/blob/c13fea90595cb1489046135c45c92d3bacb85818/core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala#L843

in that case, we'd have troubles to know the exact task cores used. We might need to maintain an extra constructor at the driver to track the resources used by the task if we want to do this alternative way.


> Instead of passing it from executor ?

The current way is actually consistent with the way we handled with custom resources, which is also assigned at driver and returned with `StatusUpdate`. Though, the custom resources also has the same issue with task cores as you concerned.

**Comment by ivoson** [comment]
Thanks for review. @Ngone51 @tgravescs @dongjoon-hyun @WeichenXu123 @mridulm

**Comment by mridulm** [comment]
Thanks @Ngone51, that is an excellent case where the approach I asked about wont work !

**Pull request closed without merging**